diff --git b/finsy/proto/gnmi1/gnmi_pb2.py a/finsy/proto/gnmi1/gnmi_pb2.py
index b1abd23..a596421 100644
--- b/finsy/proto/gnmi1/gnmi_pb2.py
+++ a/finsy/proto/gnmi1/gnmi_pb2.py
@@ -26,3 +26,3 @@ from google.protobuf import any_pb2 as google_dot_protobuf_dot_any__pb2
 from google.protobuf import descriptor_pb2 as google_dot_protobuf_dot_descriptor__pb2
-from gnmi1 import gnmi_ext_pb2 as gnmi1_dot_gnmi__ext__pb2
+from . import gnmi_ext_pb2 as gnmi1_dot_gnmi__ext__pb2
 
diff --git b/finsy/proto/gnmi1/gnmi_pb2.pyi a/finsy/proto/gnmi1/gnmi_pb2.pyi
index ef4a518..84117b1 100644
--- b/finsy/proto/gnmi1/gnmi_pb2.pyi
+++ a/finsy/proto/gnmi1/gnmi_pb2.pyi
@@ -15,3 +15,3 @@ import builtins
 import collections.abc
-import gnmi1.gnmi_ext_pb2
+from . import gnmi_ext_pb2 as _dot_gnmi_ext_pb2
 import google.protobuf.any_pb2
@@ -486,3 +486,3 @@ class SubscribeRequest(google.protobuf.message.Message):
     @property
-    def extension(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[gnmi1.gnmi_ext_pb2.Extension]:
+    def extension(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[_dot_gnmi_ext_pb2.Extension]:
         """Extension messages associated with the SubscribeRequest. See the
@@ -496,3 +496,3 @@ class SubscribeRequest(google.protobuf.message.Message):
         poll: global___Poll | None = ...,
-        extension: collections.abc.Iterable[gnmi1.gnmi_ext_pb2.Extension] | None = ...,
+        extension: collections.abc.Iterable[_dot_gnmi_ext_pb2.Extension] | None = ...,
     ) -> None: ...
@@ -549,3 +549,3 @@ class SubscribeResponse(google.protobuf.message.Message):
     @property
-    def extension(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[gnmi1.gnmi_ext_pb2.Extension]:
+    def extension(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[_dot_gnmi_ext_pb2.Extension]:
         """Extension messages associated with the SubscribeResponse. See the
@@ -560,3 +560,3 @@ class SubscribeResponse(google.protobuf.message.Message):
         error: global___Error | None = ...,
-        extension: collections.abc.Iterable[gnmi1.gnmi_ext_pb2.Extension] | None = ...,
+        extension: collections.abc.Iterable[_dot_gnmi_ext_pb2.Extension] | None = ...,
     ) -> None: ...
@@ -778,3 +778,3 @@ class SetRequest(google.protobuf.message.Message):
     @property
-    def extension(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[gnmi1.gnmi_ext_pb2.Extension]:
+    def extension(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[_dot_gnmi_ext_pb2.Extension]:
         """Extension messages associated with the SetRequest. See the
@@ -791,3 +791,3 @@ class SetRequest(google.protobuf.message.Message):
         union_replace: collections.abc.Iterable[global___Update] | None = ...,
-        extension: collections.abc.Iterable[gnmi1.gnmi_ext_pb2.Extension] | None = ...,
+        extension: collections.abc.Iterable[_dot_gnmi_ext_pb2.Extension] | None = ...,
     ) -> None: ...
@@ -833,3 +833,3 @@ class SetResponse(google.protobuf.message.Message):
     @property
-    def extension(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[gnmi1.gnmi_ext_pb2.Extension]:
+    def extension(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[_dot_gnmi_ext_pb2.Extension]:
         """Extension messages associated with the SetResponse. See the
@@ -845,3 +845,3 @@ class SetResponse(google.protobuf.message.Message):
         timestamp: builtins.int = ...,
-        extension: collections.abc.Iterable[gnmi1.gnmi_ext_pb2.Extension] | None = ...,
+        extension: collections.abc.Iterable[_dot_gnmi_ext_pb2.Extension] | None = ...,
     ) -> None: ...
@@ -991,3 +991,3 @@ class GetRequest(google.protobuf.message.Message):
     @property
-    def extension(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[gnmi1.gnmi_ext_pb2.Extension]:
+    def extension(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[_dot_gnmi_ext_pb2.Extension]:
         """Extension messages associated with the GetRequest. See the
@@ -1004,3 +1004,3 @@ class GetRequest(google.protobuf.message.Message):
         use_models: collections.abc.Iterable[global___ModelData] | None = ...,
-        extension: collections.abc.Iterable[gnmi1.gnmi_ext_pb2.Extension] | None = ...,
+        extension: collections.abc.Iterable[_dot_gnmi_ext_pb2.Extension] | None = ...,
     ) -> None: ...
@@ -1033,3 +1033,3 @@ class GetResponse(google.protobuf.message.Message):
     @property
-    def extension(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[gnmi1.gnmi_ext_pb2.Extension]:
+    def extension(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[_dot_gnmi_ext_pb2.Extension]:
         """Extension messages associated with the GetResponse. See the
@@ -1043,3 +1043,3 @@ class GetResponse(google.protobuf.message.Message):
         error: global___Error | None = ...,
-        extension: collections.abc.Iterable[gnmi1.gnmi_ext_pb2.Extension] | None = ...,
+        extension: collections.abc.Iterable[_dot_gnmi_ext_pb2.Extension] | None = ...,
     ) -> None: ...
@@ -1061,3 +1061,3 @@ class CapabilityRequest(google.protobuf.message.Message):
     @property
-    def extension(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[gnmi1.gnmi_ext_pb2.Extension]:
+    def extension(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[_dot_gnmi_ext_pb2.Extension]:
         """Extension messages associated with the CapabilityRequest. See the
@@ -1069,3 +1069,3 @@ class CapabilityRequest(google.protobuf.message.Message):
         *,
-        extension: collections.abc.Iterable[gnmi1.gnmi_ext_pb2.Extension] | None = ...,
+        extension: collections.abc.Iterable[_dot_gnmi_ext_pb2.Extension] | None = ...,
     ) -> None: ...
@@ -1099,3 +1099,3 @@ class CapabilityResponse(google.protobuf.message.Message):
     @property
-    def extension(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[gnmi1.gnmi_ext_pb2.Extension]:
+    def extension(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[_dot_gnmi_ext_pb2.Extension]:
         """Extension messages associated with the CapabilityResponse. See the
@@ -1110,3 +1110,3 @@ class CapabilityResponse(google.protobuf.message.Message):
         gNMI_version: builtins.str = ...,
-        extension: collections.abc.Iterable[gnmi1.gnmi_ext_pb2.Extension] | None = ...,
+        extension: collections.abc.Iterable[_dot_gnmi_ext_pb2.Extension] | None = ...,
     ) -> None: ...
diff --git b/finsy/proto/gnmi1/gnmi_pb2_grpc.py a/finsy/proto/gnmi1/gnmi_pb2_grpc.py
index a831028..85c8a8c 100644
--- b/finsy/proto/gnmi1/gnmi_pb2_grpc.py
+++ a/finsy/proto/gnmi1/gnmi_pb2_grpc.py
@@ -5,3 +5,3 @@ import warnings
 
-from gnmi1 import gnmi_pb2 as gnmi1_dot_gnmi__pb2
+from . import gnmi_pb2 as gnmi1_dot_gnmi__pb2
 
diff --git b/finsy/proto/gnmi1/gnmi_pb2_grpc.pyi a/finsy/proto/gnmi1/gnmi_pb2_grpc.pyi
index cbb2816..c1a2cef 100644
--- b/finsy/proto/gnmi1/gnmi_pb2_grpc.pyi
+++ a/finsy/proto/gnmi1/gnmi_pb2_grpc.pyi
@@ -15,3 +15,3 @@ import abc
 import collections.abc
-import gnmi1.gnmi_pb2
+from . import gnmi_pb2 as _dot_gnmi_pb2
 import grpc
@@ -30,4 +30,4 @@ class gNMIStub:
     Capabilities: grpc.UnaryUnaryMultiCallable[
-        gnmi1.gnmi_pb2.CapabilityRequest,
-        gnmi1.gnmi_pb2.CapabilityResponse,
+        _dot_gnmi_pb2.CapabilityRequest,
+        _dot_gnmi_pb2.CapabilityResponse,
     ]
@@ -42,4 +42,4 @@ class gNMIStub:
     Get: grpc.UnaryUnaryMultiCallable[
-        gnmi1.gnmi_pb2.GetRequest,
-        gnmi1.gnmi_pb2.GetResponse,
+        _dot_gnmi_pb2.GetRequest,
+        _dot_gnmi_pb2.GetResponse,
     ]
@@ -53,4 +53,4 @@ class gNMIStub:
     Set: grpc.UnaryUnaryMultiCallable[
-        gnmi1.gnmi_pb2.SetRequest,
-        gnmi1.gnmi_pb2.SetResponse,
+        _dot_gnmi_pb2.SetRequest,
+        _dot_gnmi_pb2.SetResponse,
     ]
@@ -63,4 +63,4 @@ class gNMIStub:
     Subscribe: grpc.StreamStreamMultiCallable[
-        gnmi1.gnmi_pb2.SubscribeRequest,
-        gnmi1.gnmi_pb2.SubscribeResponse,
+        _dot_gnmi_pb2.SubscribeRequest,
+        _dot_gnmi_pb2.SubscribeResponse,
     ]
@@ -75,4 +75,4 @@ class gNMIAsyncStub:
     Capabilities: grpc.aio.UnaryUnaryMultiCallable[
-        gnmi1.gnmi_pb2.CapabilityRequest,
-        gnmi1.gnmi_pb2.CapabilityResponse,
+        _dot_gnmi_pb2.CapabilityRequest,
+        _dot_gnmi_pb2.CapabilityResponse,
     ]
@@ -87,4 +87,4 @@ class gNMIAsyncStub:
     Get: grpc.aio.UnaryUnaryMultiCallable[
-        gnmi1.gnmi_pb2.GetRequest,
-        gnmi1.gnmi_pb2.GetResponse,
+        _dot_gnmi_pb2.GetRequest,
+        _dot_gnmi_pb2.GetResponse,
     ]
@@ -98,4 +98,4 @@ class gNMIAsyncStub:
     Set: grpc.aio.UnaryUnaryMultiCallable[
-        gnmi1.gnmi_pb2.SetRequest,
-        gnmi1.gnmi_pb2.SetResponse,
+        _dot_gnmi_pb2.SetRequest,
+        _dot_gnmi_pb2.SetResponse,
     ]
@@ -108,4 +108,4 @@ class gNMIAsyncStub:
     Subscribe: grpc.aio.StreamStreamMultiCallable[
-        gnmi1.gnmi_pb2.SubscribeRequest,
-        gnmi1.gnmi_pb2.SubscribeResponse,
+        _dot_gnmi_pb2.SubscribeRequest,
+        _dot_gnmi_pb2.SubscribeResponse,
     ]
@@ -122,5 +122,5 @@ class gNMIServicer(metaclass=abc.ABCMeta):
         self,
-        request: gnmi1.gnmi_pb2.CapabilityRequest,
+        request: _dot_gnmi_pb2.CapabilityRequest,
         context: _ServicerContext,
-    ) -> typing.Union[gnmi1.gnmi_pb2.CapabilityResponse, collections.abc.Awaitable[gnmi1.gnmi_pb2.CapabilityResponse]]:
+    ) -> typing.Union[_dot_gnmi_pb2.CapabilityResponse, collections.abc.Awaitable[_dot_gnmi_pb2.CapabilityResponse]]:
         """Capabilities allows the client to retrieve the set of capabilities that
@@ -136,5 +136,5 @@ class gNMIServicer(metaclass=abc.ABCMeta):
         self,
-        request: gnmi1.gnmi_pb2.GetRequest,
+        request: _dot_gnmi_pb2.GetRequest,
         context: _ServicerContext,
-    ) -> typing.Union[gnmi1.gnmi_pb2.GetResponse, collections.abc.Awaitable[gnmi1.gnmi_pb2.GetResponse]]:
+    ) -> typing.Union[_dot_gnmi_pb2.GetResponse, collections.abc.Awaitable[_dot_gnmi_pb2.GetResponse]]:
         """Retrieve a snapshot of data from the target. A Get RPC requests that the
@@ -149,5 +149,5 @@ class gNMIServicer(metaclass=abc.ABCMeta):
         self,
-        request: gnmi1.gnmi_pb2.SetRequest,
+        request: _dot_gnmi_pb2.SetRequest,
         context: _ServicerContext,
-    ) -> typing.Union[gnmi1.gnmi_pb2.SetResponse, collections.abc.Awaitable[gnmi1.gnmi_pb2.SetResponse]]:
+    ) -> typing.Union[_dot_gnmi_pb2.SetResponse, collections.abc.Awaitable[_dot_gnmi_pb2.SetResponse]]:
         """Set allows the client to modify the state of data on the target. The
@@ -161,5 +161,5 @@ class gNMIServicer(metaclass=abc.ABCMeta):
         self,
-        request_iterator: _MaybeAsyncIterator[gnmi1.gnmi_pb2.SubscribeRequest],
+        request_iterator: _MaybeAsyncIterator[_dot_gnmi_pb2.SubscribeRequest],
         context: _ServicerContext,
-    ) -> typing.Union[collections.abc.Iterator[gnmi1.gnmi_pb2.SubscribeResponse], collections.abc.AsyncIterator[gnmi1.gnmi_pb2.SubscribeResponse]]:
+    ) -> typing.Union[collections.abc.Iterator[_dot_gnmi_pb2.SubscribeResponse], collections.abc.AsyncIterator[_dot_gnmi_pb2.SubscribeResponse]]:
         """Subscribe allows a client to request the target to send it values
diff --git b/finsy/proto/p4/config/v1/p4info_pb2.py a/finsy/proto/p4/config/v1/p4info_pb2.py
index 99d4db5..1d461db 100644
--- b/finsy/proto/p4/config/v1/p4info_pb2.py
+++ a/finsy/proto/p4/config/v1/p4info_pb2.py
@@ -25,3 +25,3 @@ _sym_db = _symbol_database.Default()
 from google.protobuf import any_pb2 as google_dot_protobuf_dot_any__pb2
-from p4.config.v1 import p4types_pb2 as p4_dot_config_dot_v1_dot_p4types__pb2
+from . import p4types_pb2 as p4_dot_config_dot_v1_dot_p4types__pb2
 
diff --git b/finsy/proto/p4/config/v1/p4info_pb2.pyi a/finsy/proto/p4/config/v1/p4info_pb2.pyi
index ebd8515..3e171a6 100644
--- b/finsy/proto/p4/config/v1/p4info_pb2.pyi
+++ a/finsy/proto/p4/config/v1/p4info_pb2.pyi
@@ -12,3 +12,3 @@ import google.protobuf.internal.enum_type_wrapper
 import google.protobuf.message
-import p4.config.v1.p4types_pb2
+from . import p4types_pb2 as _dot_p4types_pb2
 import sys
@@ -68,3 +68,3 @@ class P4Info(google.protobuf.message.Message):
     @property
-    def type_info(self) -> p4.config.v1.p4types_pb2.P4TypeInfo: ...
+    def type_info(self) -> _dot_p4types_pb2.P4TypeInfo: ...
     def __init__(
@@ -85,3 +85,3 @@ class P4Info(google.protobuf.message.Message):
         externs: collections.abc.Iterable[global___Extern] | None = ...,
-        type_info: p4.config.v1.p4types_pb2.P4TypeInfo | None = ...,
+        type_info: _dot_p4types_pb2.P4TypeInfo | None = ...,
     ) -> None: ...
@@ -191,3 +191,3 @@ class PkgInfo(google.protobuf.message.Message):
     @property
-    def annotation_locations(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[p4.config.v1.p4types_pb2.SourceLocation]:
+    def annotation_locations(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[_dot_p4types_pb2.SourceLocation]:
         """Optional. If present, the location of `annotations[i]` is given by
@@ -197,3 +197,3 @@ class PkgInfo(google.protobuf.message.Message):
     @property
-    def structured_annotations(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[p4.config.v1.p4types_pb2.StructuredAnnotation]:
+    def structured_annotations(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[_dot_p4types_pb2.StructuredAnnotation]:
         """Miscellaneous metadata, structured; a way to extend PkgInfo"""
@@ -215,3 +215,3 @@ class PkgInfo(google.protobuf.message.Message):
         annotations: collections.abc.Iterable[builtins.str] | None = ...,
-        annotation_locations: collections.abc.Iterable[p4.config.v1.p4types_pb2.SourceLocation] | None = ...,
+        annotation_locations: collections.abc.Iterable[_dot_p4types_pb2.SourceLocation] | None = ...,
         arch: builtins.str = ...,
@@ -220,3 +220,3 @@ class PkgInfo(google.protobuf.message.Message):
         url: builtins.str = ...,
-        structured_annotations: collections.abc.Iterable[p4.config.v1.p4types_pb2.StructuredAnnotation] | None = ...,
+        structured_annotations: collections.abc.Iterable[_dot_p4types_pb2.StructuredAnnotation] | None = ...,
         platform_properties: global___PlatformProperties | None = ...,
@@ -340,3 +340,3 @@ class Preamble(google.protobuf.message.Message):
     @property
-    def annotation_locations(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[p4.config.v1.p4types_pb2.SourceLocation]:
+    def annotation_locations(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[_dot_p4types_pb2.SourceLocation]:
         """Optional. If present, the location of `annotations[i]` is given by
@@ -350,3 +350,3 @@ class Preamble(google.protobuf.message.Message):
     @property
-    def structured_annotations(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[p4.config.v1.p4types_pb2.StructuredAnnotation]: ...
+    def structured_annotations(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[_dot_p4types_pb2.StructuredAnnotation]: ...
     def __init__(
@@ -358,5 +358,5 @@ class Preamble(google.protobuf.message.Message):
         annotations: collections.abc.Iterable[builtins.str] | None = ...,
-        annotation_locations: collections.abc.Iterable[p4.config.v1.p4types_pb2.SourceLocation] | None = ...,
+        annotation_locations: collections.abc.Iterable[_dot_p4types_pb2.SourceLocation] | None = ...,
         doc: global___Documentation | None = ...,
-        structured_annotations: collections.abc.Iterable[p4.config.v1.p4types_pb2.StructuredAnnotation] | None = ...,
+        structured_annotations: collections.abc.Iterable[_dot_p4types_pb2.StructuredAnnotation] | None = ...,
     ) -> None: ...
@@ -465,3 +465,3 @@ class MatchField(google.protobuf.message.Message):
     @property
-    def annotation_locations(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[p4.config.v1.p4types_pb2.SourceLocation]:
+    def annotation_locations(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[_dot_p4types_pb2.SourceLocation]:
         """Optional. If present, the location of `annotations[i]` is given by
@@ -475,3 +475,3 @@ class MatchField(google.protobuf.message.Message):
     @property
-    def type_name(self) -> p4.config.v1.p4types_pb2.P4NamedType:
+    def type_name(self) -> _dot_p4types_pb2.P4NamedType:
         """unset if not user-defined type"""
@@ -479,3 +479,3 @@ class MatchField(google.protobuf.message.Message):
     @property
-    def structured_annotations(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[p4.config.v1.p4types_pb2.StructuredAnnotation]: ...
+    def structured_annotations(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[_dot_p4types_pb2.StructuredAnnotation]: ...
     def __init__(
@@ -486,3 +486,3 @@ class MatchField(google.protobuf.message.Message):
         annotations: collections.abc.Iterable[builtins.str] | None = ...,
-        annotation_locations: collections.abc.Iterable[p4.config.v1.p4types_pb2.SourceLocation] | None = ...,
+        annotation_locations: collections.abc.Iterable[_dot_p4types_pb2.SourceLocation] | None = ...,
         bitwidth: builtins.int = ...,
@@ -491,4 +491,4 @@ class MatchField(google.protobuf.message.Message):
         doc: global___Documentation | None = ...,
-        type_name: p4.config.v1.p4types_pb2.P4NamedType | None = ...,
-        structured_annotations: collections.abc.Iterable[p4.config.v1.p4types_pb2.StructuredAnnotation] | None = ...,
+        type_name: _dot_p4types_pb2.P4NamedType | None = ...,
+        structured_annotations: collections.abc.Iterable[_dot_p4types_pb2.StructuredAnnotation] | None = ...,
     ) -> None: ...
@@ -684,3 +684,3 @@ class ActionRef(google.protobuf.message.Message):
     @property
-    def annotation_locations(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[p4.config.v1.p4types_pb2.SourceLocation]:
+    def annotation_locations(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[_dot_p4types_pb2.SourceLocation]:
         """Optional. If present, the location of `annotations[i]` is given by
@@ -690,3 +690,3 @@ class ActionRef(google.protobuf.message.Message):
     @property
-    def structured_annotations(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[p4.config.v1.p4types_pb2.StructuredAnnotation]: ...
+    def structured_annotations(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[_dot_p4types_pb2.StructuredAnnotation]: ...
     def __init__(
@@ -697,4 +697,4 @@ class ActionRef(google.protobuf.message.Message):
         annotations: collections.abc.Iterable[builtins.str] | None = ...,
-        annotation_locations: collections.abc.Iterable[p4.config.v1.p4types_pb2.SourceLocation] | None = ...,
-        structured_annotations: collections.abc.Iterable[p4.config.v1.p4types_pb2.StructuredAnnotation] | None = ...,
+        annotation_locations: collections.abc.Iterable[_dot_p4types_pb2.SourceLocation] | None = ...,
+        structured_annotations: collections.abc.Iterable[_dot_p4types_pb2.StructuredAnnotation] | None = ...,
     ) -> None: ...
@@ -726,3 +726,3 @@ class Action(google.protobuf.message.Message):
         @property
-        def annotation_locations(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[p4.config.v1.p4types_pb2.SourceLocation]:
+        def annotation_locations(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[_dot_p4types_pb2.SourceLocation]:
             """Optional. If present, the location of `annotations[i]` is given by
@@ -736,3 +736,3 @@ class Action(google.protobuf.message.Message):
         @property
-        def type_name(self) -> p4.config.v1.p4types_pb2.P4NamedType:
+        def type_name(self) -> _dot_p4types_pb2.P4NamedType:
             """unset if not user-defined type"""
@@ -740,3 +740,3 @@ class Action(google.protobuf.message.Message):
         @property
-        def structured_annotations(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[p4.config.v1.p4types_pb2.StructuredAnnotation]: ...
+        def structured_annotations(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[_dot_p4types_pb2.StructuredAnnotation]: ...
         def __init__(
@@ -747,7 +747,7 @@ class Action(google.protobuf.message.Message):
             annotations: collections.abc.Iterable[builtins.str] | None = ...,
-            annotation_locations: collections.abc.Iterable[p4.config.v1.p4types_pb2.SourceLocation] | None = ...,
+            annotation_locations: collections.abc.Iterable[_dot_p4types_pb2.SourceLocation] | None = ...,
             bitwidth: builtins.int = ...,
             doc: global___Documentation | None = ...,
-            type_name: p4.config.v1.p4types_pb2.P4NamedType | None = ...,
-            structured_annotations: collections.abc.Iterable[p4.config.v1.p4types_pb2.StructuredAnnotation] | None = ...,
+            type_name: _dot_p4types_pb2.P4NamedType | None = ...,
+            structured_annotations: collections.abc.Iterable[_dot_p4types_pb2.StructuredAnnotation] | None = ...,
         ) -> None: ...
@@ -915,3 +915,3 @@ class Counter(google.protobuf.message.Message):
     @property
-    def index_type_name(self) -> p4.config.v1.p4types_pb2.P4NamedType:
+    def index_type_name(self) -> _dot_p4types_pb2.P4NamedType:
         """unset if index is not user-defined type"""
@@ -924,3 +924,3 @@ class Counter(google.protobuf.message.Message):
         size: builtins.int = ...,
-        index_type_name: p4.config.v1.p4types_pb2.P4NamedType | None = ...,
+        index_type_name: _dot_p4types_pb2.P4NamedType | None = ...,
     ) -> None: ...
@@ -1059,3 +1059,3 @@ class Meter(google.protobuf.message.Message):
     @property
-    def index_type_name(self) -> p4.config.v1.p4types_pb2.P4NamedType:
+    def index_type_name(self) -> _dot_p4types_pb2.P4NamedType:
         """unset if index is not user-defined type"""
@@ -1068,3 +1068,3 @@ class Meter(google.protobuf.message.Message):
         size: builtins.int = ...,
-        index_type_name: p4.config.v1.p4types_pb2.P4NamedType | None = ...,
+        index_type_name: _dot_p4types_pb2.P4NamedType | None = ...,
     ) -> None: ...
@@ -1132,3 +1132,3 @@ class ControllerPacketMetadata(google.protobuf.message.Message):
         @property
-        def annotation_locations(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[p4.config.v1.p4types_pb2.SourceLocation]:
+        def annotation_locations(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[_dot_p4types_pb2.SourceLocation]:
             """Optional. If present, the location of `annotations[i]` is given by
@@ -1138,3 +1138,3 @@ class ControllerPacketMetadata(google.protobuf.message.Message):
         @property
-        def type_name(self) -> p4.config.v1.p4types_pb2.P4NamedType:
+        def type_name(self) -> _dot_p4types_pb2.P4NamedType:
             """unset if not user-defined type"""
@@ -1142,3 +1142,3 @@ class ControllerPacketMetadata(google.protobuf.message.Message):
         @property
-        def structured_annotations(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[p4.config.v1.p4types_pb2.StructuredAnnotation]: ...
+        def structured_annotations(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[_dot_p4types_pb2.StructuredAnnotation]: ...
         def __init__(
@@ -1149,6 +1149,6 @@ class ControllerPacketMetadata(google.protobuf.message.Message):
             annotations: collections.abc.Iterable[builtins.str] | None = ...,
-            annotation_locations: collections.abc.Iterable[p4.config.v1.p4types_pb2.SourceLocation] | None = ...,
+            annotation_locations: collections.abc.Iterable[_dot_p4types_pb2.SourceLocation] | None = ...,
             bitwidth: builtins.int = ...,
-            type_name: p4.config.v1.p4types_pb2.P4NamedType | None = ...,
-            structured_annotations: collections.abc.Iterable[p4.config.v1.p4types_pb2.StructuredAnnotation] | None = ...,
+            type_name: _dot_p4types_pb2.P4NamedType | None = ...,
+            structured_annotations: collections.abc.Iterable[_dot_p4types_pb2.StructuredAnnotation] | None = ...,
         ) -> None: ...
@@ -1219,5 +1219,5 @@ class Register(google.protobuf.message.Message):
     @property
-    def type_spec(self) -> p4.config.v1.p4types_pb2.P4DataTypeSpec: ...
+    def type_spec(self) -> _dot_p4types_pb2.P4DataTypeSpec: ...
     @property
-    def index_type_name(self) -> p4.config.v1.p4types_pb2.P4NamedType:
+    def index_type_name(self) -> _dot_p4types_pb2.P4NamedType:
         """unset if index is not user-defined type"""
@@ -1228,5 +1228,5 @@ class Register(google.protobuf.message.Message):
         preamble: global___Preamble | None = ...,
-        type_spec: p4.config.v1.p4types_pb2.P4DataTypeSpec | None = ...,
+        type_spec: _dot_p4types_pb2.P4DataTypeSpec | None = ...,
         size: builtins.int = ...,
-        index_type_name: p4.config.v1.p4types_pb2.P4NamedType | None = ...,
+        index_type_name: _dot_p4types_pb2.P4NamedType | None = ...,
     ) -> None: ...
@@ -1246,3 +1246,3 @@ class Digest(google.protobuf.message.Message):
     @property
-    def type_spec(self) -> p4.config.v1.p4types_pb2.P4DataTypeSpec: ...
+    def type_spec(self) -> _dot_p4types_pb2.P4DataTypeSpec: ...
     def __init__(
@@ -1251,3 +1251,3 @@ class Digest(google.protobuf.message.Message):
         preamble: global___Preamble | None = ...,
-        type_spec: p4.config.v1.p4types_pb2.P4DataTypeSpec | None = ...,
+        type_spec: _dot_p4types_pb2.P4DataTypeSpec | None = ...,
     ) -> None: ...
diff --git b/finsy/proto/p4/v1/p4runtime_pb2.py a/finsy/proto/p4/v1/p4runtime_pb2.py
index fe1b877..5777af0 100644
--- b/finsy/proto/p4/v1/p4runtime_pb2.py
+++ a/finsy/proto/p4/v1/p4runtime_pb2.py
@@ -25,5 +25,5 @@ _sym_db = _symbol_database.Default()
 from google.protobuf import any_pb2 as google_dot_protobuf_dot_any__pb2
-from google.rpc import status_pb2 as google_dot_rpc_dot_status__pb2
-from p4.config.v1 import p4info_pb2 as p4_dot_config_dot_v1_dot_p4info__pb2
-from p4.v1 import p4data_pb2 as p4_dot_v1_dot_p4data__pb2
+from ...google.rpc import status_pb2 as google_dot_rpc_dot_status__pb2
+from ..config.v1 import p4info_pb2 as p4_dot_config_dot_v1_dot_p4info__pb2
+from . import p4data_pb2 as p4_dot_v1_dot_p4data__pb2
 
diff --git b/finsy/proto/p4/v1/p4runtime_pb2.pyi a/finsy/proto/p4/v1/p4runtime_pb2.pyi
index d369349..903e635 100644
--- b/finsy/proto/p4/v1/p4runtime_pb2.pyi
+++ a/finsy/proto/p4/v1/p4runtime_pb2.pyi
@@ -12,5 +12,5 @@ import google.protobuf.internal.enum_type_wrapper
 import google.protobuf.message
-import google.rpc.status_pb2
-import p4.config.v1.p4info_pb2
-import p4.v1.p4data_pb2
+from ...google.rpc import status_pb2 as _dot_status_pb2
+from ..config.v1 import p4info_pb2 as _dot_p4info_pb2
+from . import p4data_pb2 as _dot_p4data_pb2
 import sys
@@ -1274,3 +1274,3 @@ class RegisterEntry(google.protobuf.message.Message):
     @property
-    def data(self) -> p4.v1.p4data_pb2.P4Data: ...
+    def data(self) -> _dot_p4data_pb2.P4Data: ...
     def __init__(
@@ -1280,3 +1280,3 @@ class RegisterEntry(google.protobuf.message.Message):
         index: global___Index | None = ...,
-        data: p4.v1.p4data_pb2.P4Data | None = ...,
+        data: _dot_p4data_pb2.P4Data | None = ...,
     ) -> None: ...
@@ -1507,3 +1507,3 @@ class DigestList(google.protobuf.message.Message):
     @property
-    def data(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[p4.v1.p4data_pb2.P4Data]:
+    def data(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[_dot_p4data_pb2.P4Data]:
         """List of entries: each call to the Digest<T>::pack() method corresponds to
@@ -1517,3 +1517,3 @@ class DigestList(google.protobuf.message.Message):
         list_id: builtins.int = ...,
-        data: collections.abc.Iterable[p4.v1.p4data_pb2.P4Data] | None = ...,
+        data: collections.abc.Iterable[_dot_p4data_pb2.P4Data] | None = ...,
         timestamp: builtins.int = ...,
@@ -1575,3 +1575,3 @@ class MasterArbitrationUpdate(google.protobuf.message.Message):
     @property
-    def status(self) -> google.rpc.status_pb2.Status:
+    def status(self) -> _dot_status_pb2.Status:
         """Switch populates this with OK for the client that is the primary, and
@@ -1587,3 +1587,3 @@ class MasterArbitrationUpdate(google.protobuf.message.Message):
         election_id: global___Uint128 | None = ...,
-        status: google.rpc.status_pb2.Status | None = ...,
+        status: _dot_status_pb2.Status | None = ...,
     ) -> None: ...
@@ -1944,3 +1944,3 @@ class ForwardingPipelineConfig(google.protobuf.message.Message):
     @property
-    def p4info(self) -> p4.config.v1.p4info_pb2.P4Info: ...
+    def p4info(self) -> _dot_p4info_pb2.P4Info: ...
     @property
@@ -1950,3 +1950,3 @@ class ForwardingPipelineConfig(google.protobuf.message.Message):
         *,
-        p4info: p4.config.v1.p4info_pb2.P4Info | None = ...,
+        p4info: _dot_p4info_pb2.P4Info | None = ...,
         p4_device_config: builtins.bytes = ...,
diff --git b/finsy/proto/p4/v1/p4runtime_pb2_grpc.py a/finsy/proto/p4/v1/p4runtime_pb2_grpc.py
index 51c39b2..a1dfc5c 100644
--- b/finsy/proto/p4/v1/p4runtime_pb2_grpc.py
+++ a/finsy/proto/p4/v1/p4runtime_pb2_grpc.py
@@ -5,3 +5,3 @@ import warnings
 
-from p4.v1 import p4runtime_pb2 as p4_dot_v1_dot_p4runtime__pb2
+from . import p4runtime_pb2 as p4_dot_v1_dot_p4runtime__pb2
 
diff --git b/finsy/proto/p4/v1/p4runtime_pb2_grpc.pyi a/finsy/proto/p4/v1/p4runtime_pb2_grpc.pyi
index bfe8ee1..925392e 100644
--- b/finsy/proto/p4/v1/p4runtime_pb2_grpc.pyi
+++ a/finsy/proto/p4/v1/p4runtime_pb2_grpc.pyi
@@ -9,3 +9,3 @@ import grpc
 import grpc.aio
-import p4.v1.p4runtime_pb2
+from . import p4runtime_pb2 as _dot_p4runtime_pb2
 import typing
@@ -22,4 +22,4 @@ class P4RuntimeStub:
     Write: grpc.UnaryUnaryMultiCallable[
-        p4.v1.p4runtime_pb2.WriteRequest,
-        p4.v1.p4runtime_pb2.WriteResponse,
+        _dot_p4runtime_pb2.WriteRequest,
+        _dot_p4runtime_pb2.WriteResponse,
     ]
@@ -28,4 +28,4 @@ class P4RuntimeStub:
     Read: grpc.UnaryStreamMultiCallable[
-        p4.v1.p4runtime_pb2.ReadRequest,
-        p4.v1.p4runtime_pb2.ReadResponse,
+        _dot_p4runtime_pb2.ReadRequest,
+        _dot_p4runtime_pb2.ReadResponse,
     ]
@@ -34,4 +34,4 @@ class P4RuntimeStub:
     SetForwardingPipelineConfig: grpc.UnaryUnaryMultiCallable[
-        p4.v1.p4runtime_pb2.SetForwardingPipelineConfigRequest,
-        p4.v1.p4runtime_pb2.SetForwardingPipelineConfigResponse,
+        _dot_p4runtime_pb2.SetForwardingPipelineConfigRequest,
+        _dot_p4runtime_pb2.SetForwardingPipelineConfigResponse,
     ]
@@ -40,4 +40,4 @@ class P4RuntimeStub:
     GetForwardingPipelineConfig: grpc.UnaryUnaryMultiCallable[
-        p4.v1.p4runtime_pb2.GetForwardingPipelineConfigRequest,
-        p4.v1.p4runtime_pb2.GetForwardingPipelineConfigResponse,
+        _dot_p4runtime_pb2.GetForwardingPipelineConfigRequest,
+        _dot_p4runtime_pb2.GetForwardingPipelineConfigResponse,
     ]
@@ -46,4 +46,4 @@ class P4RuntimeStub:
     StreamChannel: grpc.StreamStreamMultiCallable[
-        p4.v1.p4runtime_pb2.StreamMessageRequest,
-        p4.v1.p4runtime_pb2.StreamMessageResponse,
+        _dot_p4runtime_pb2.StreamMessageRequest,
+        _dot_p4runtime_pb2.StreamMessageResponse,
     ]
@@ -62,4 +62,4 @@ class P4RuntimeStub:
     Capabilities: grpc.UnaryUnaryMultiCallable[
-        p4.v1.p4runtime_pb2.CapabilitiesRequest,
-        p4.v1.p4runtime_pb2.CapabilitiesResponse,
+        _dot_p4runtime_pb2.CapabilitiesRequest,
+        _dot_p4runtime_pb2.CapabilitiesResponse,
     ]
@@ -68,4 +68,4 @@ class P4RuntimeAsyncStub:
     Write: grpc.aio.UnaryUnaryMultiCallable[
-        p4.v1.p4runtime_pb2.WriteRequest,
-        p4.v1.p4runtime_pb2.WriteResponse,
+        _dot_p4runtime_pb2.WriteRequest,
+        _dot_p4runtime_pb2.WriteResponse,
     ]
@@ -74,4 +74,4 @@ class P4RuntimeAsyncStub:
     Read: grpc.aio.UnaryStreamMultiCallable[
-        p4.v1.p4runtime_pb2.ReadRequest,
-        p4.v1.p4runtime_pb2.ReadResponse,
+        _dot_p4runtime_pb2.ReadRequest,
+        _dot_p4runtime_pb2.ReadResponse,
     ]
@@ -80,4 +80,4 @@ class P4RuntimeAsyncStub:
     SetForwardingPipelineConfig: grpc.aio.UnaryUnaryMultiCallable[
-        p4.v1.p4runtime_pb2.SetForwardingPipelineConfigRequest,
-        p4.v1.p4runtime_pb2.SetForwardingPipelineConfigResponse,
+        _dot_p4runtime_pb2.SetForwardingPipelineConfigRequest,
+        _dot_p4runtime_pb2.SetForwardingPipelineConfigResponse,
     ]
@@ -86,4 +86,4 @@ class P4RuntimeAsyncStub:
     GetForwardingPipelineConfig: grpc.aio.UnaryUnaryMultiCallable[
-        p4.v1.p4runtime_pb2.GetForwardingPipelineConfigRequest,
-        p4.v1.p4runtime_pb2.GetForwardingPipelineConfigResponse,
+        _dot_p4runtime_pb2.GetForwardingPipelineConfigRequest,
+        _dot_p4runtime_pb2.GetForwardingPipelineConfigResponse,
     ]
@@ -92,4 +92,4 @@ class P4RuntimeAsyncStub:
     StreamChannel: grpc.aio.StreamStreamMultiCallable[
-        p4.v1.p4runtime_pb2.StreamMessageRequest,
-        p4.v1.p4runtime_pb2.StreamMessageResponse,
+        _dot_p4runtime_pb2.StreamMessageRequest,
+        _dot_p4runtime_pb2.StreamMessageResponse,
     ]
@@ -108,4 +108,4 @@ class P4RuntimeAsyncStub:
     Capabilities: grpc.aio.UnaryUnaryMultiCallable[
-        p4.v1.p4runtime_pb2.CapabilitiesRequest,
-        p4.v1.p4runtime_pb2.CapabilitiesResponse,
+        _dot_p4runtime_pb2.CapabilitiesRequest,
+        _dot_p4runtime_pb2.CapabilitiesResponse,
     ]
@@ -116,5 +116,5 @@ class P4RuntimeServicer(metaclass=abc.ABCMeta):
         self,
-        request: p4.v1.p4runtime_pb2.WriteRequest,
+        request: _dot_p4runtime_pb2.WriteRequest,
         context: _ServicerContext,
-    ) -> typing.Union[p4.v1.p4runtime_pb2.WriteResponse, collections.abc.Awaitable[p4.v1.p4runtime_pb2.WriteResponse]]:
+    ) -> typing.Union[_dot_p4runtime_pb2.WriteResponse, collections.abc.Awaitable[_dot_p4runtime_pb2.WriteResponse]]:
         """Update one or more P4 entities on the target."""
@@ -124,5 +124,5 @@ class P4RuntimeServicer(metaclass=abc.ABCMeta):
         self,
-        request: p4.v1.p4runtime_pb2.ReadRequest,
+        request: _dot_p4runtime_pb2.ReadRequest,
         context: _ServicerContext,
-    ) -> typing.Union[collections.abc.Iterator[p4.v1.p4runtime_pb2.ReadResponse], collections.abc.AsyncIterator[p4.v1.p4runtime_pb2.ReadResponse]]:
+    ) -> typing.Union[collections.abc.Iterator[_dot_p4runtime_pb2.ReadResponse], collections.abc.AsyncIterator[_dot_p4runtime_pb2.ReadResponse]]:
         """Read one or more P4 entities from the target."""
@@ -132,5 +132,5 @@ class P4RuntimeServicer(metaclass=abc.ABCMeta):
         self,
-        request: p4.v1.p4runtime_pb2.SetForwardingPipelineConfigRequest,
+        request: _dot_p4runtime_pb2.SetForwardingPipelineConfigRequest,
         context: _ServicerContext,
-    ) -> typing.Union[p4.v1.p4runtime_pb2.SetForwardingPipelineConfigResponse, collections.abc.Awaitable[p4.v1.p4runtime_pb2.SetForwardingPipelineConfigResponse]]:
+    ) -> typing.Union[_dot_p4runtime_pb2.SetForwardingPipelineConfigResponse, collections.abc.Awaitable[_dot_p4runtime_pb2.SetForwardingPipelineConfigResponse]]:
         """Sets the P4 forwarding-pipeline config."""
@@ -140,5 +140,5 @@ class P4RuntimeServicer(metaclass=abc.ABCMeta):
         self,
-        request: p4.v1.p4runtime_pb2.GetForwardingPipelineConfigRequest,
+        request: _dot_p4runtime_pb2.GetForwardingPipelineConfigRequest,
         context: _ServicerContext,
-    ) -> typing.Union[p4.v1.p4runtime_pb2.GetForwardingPipelineConfigResponse, collections.abc.Awaitable[p4.v1.p4runtime_pb2.GetForwardingPipelineConfigResponse]]:
+    ) -> typing.Union[_dot_p4runtime_pb2.GetForwardingPipelineConfigResponse, collections.abc.Awaitable[_dot_p4runtime_pb2.GetForwardingPipelineConfigResponse]]:
         """Gets the current P4 forwarding-pipeline config."""
@@ -148,5 +148,5 @@ class P4RuntimeServicer(metaclass=abc.ABCMeta):
         self,
-        request_iterator: _MaybeAsyncIterator[p4.v1.p4runtime_pb2.StreamMessageRequest],
+        request_iterator: _MaybeAsyncIterator[_dot_p4runtime_pb2.StreamMessageRequest],
         context: _ServicerContext,
-    ) -> typing.Union[collections.abc.Iterator[p4.v1.p4runtime_pb2.StreamMessageResponse], collections.abc.AsyncIterator[p4.v1.p4runtime_pb2.StreamMessageResponse]]:
+    ) -> typing.Union[collections.abc.Iterator[_dot_p4runtime_pb2.StreamMessageResponse], collections.abc.AsyncIterator[_dot_p4runtime_pb2.StreamMessageResponse]]:
         """Represents the bidirectional stream between the controller and the
@@ -166,5 +166,5 @@ class P4RuntimeServicer(metaclass=abc.ABCMeta):
         self,
-        request: p4.v1.p4runtime_pb2.CapabilitiesRequest,
+        request: _dot_p4runtime_pb2.CapabilitiesRequest,
         context: _ServicerContext,
-    ) -> typing.Union[p4.v1.p4runtime_pb2.CapabilitiesResponse, collections.abc.Awaitable[p4.v1.p4runtime_pb2.CapabilitiesResponse]]: ...
+    ) -> typing.Union[_dot_p4runtime_pb2.CapabilitiesResponse, collections.abc.Awaitable[_dot_p4runtime_pb2.CapabilitiesResponse]]: ...
 
diff --git b/finsy/proto/p4testgen1/p4testgen_pb2.py a/finsy/proto/p4testgen1/p4testgen_pb2.py
index fbfd6fc..f663c15 100644
--- b/finsy/proto/p4testgen1/p4testgen_pb2.py
+++ a/finsy/proto/p4testgen1/p4testgen_pb2.py
@@ -24,3 +24,3 @@ _sym_db = _symbol_database.Default()
 
-from p4.v1 import p4runtime_pb2 as p4_dot_v1_dot_p4runtime__pb2
+from ..p4.v1 import p4runtime_pb2 as p4_dot_v1_dot_p4runtime__pb2
 
diff --git b/finsy/proto/p4testgen1/p4testgen_pb2.pyi a/finsy/proto/p4testgen1/p4testgen_pb2.pyi
index c7d12ae..7b0770d 100644
--- b/finsy/proto/p4testgen1/p4testgen_pb2.pyi
+++ a/finsy/proto/p4testgen1/p4testgen_pb2.pyi
@@ -10,3 +10,3 @@ import google.protobuf.internal.containers
 import google.protobuf.message
-import p4.v1.p4runtime_pb2
+from ..p4.v1 import p4runtime_pb2 as _dot_p4runtime_pb2
 import typing
@@ -77,3 +77,3 @@ class TestCase(google.protobuf.message.Message):
     @property
-    def entities(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[p4.v1.p4runtime_pb2.Entity]:
+    def entities(self) -> google.protobuf.internal.containers.RepeatedCompositeFieldContainer[_dot_p4runtime_pb2.Entity]:
         """The entities (e.g., table entries) to install on the switch before
@@ -95,3 +95,3 @@ class TestCase(google.protobuf.message.Message):
         expected_output_packet: collections.abc.Iterable[global___OutputPacketAtPort] | None = ...,
-        entities: collections.abc.Iterable[p4.v1.p4runtime_pb2.Entity] | None = ...,
+        entities: collections.abc.Iterable[_dot_p4runtime_pb2.Entity] | None = ...,
         traces: collections.abc.Iterable[builtins.str] | None = ...,
