diff --git b/finsy/proto/gnmi1/gnmi_pb2.py a/finsy/proto/gnmi1/gnmi_pb2.py
index c2b36d4..f83da65 100644
--- b/finsy/proto/gnmi1/gnmi_pb2.py
+++ a/finsy/proto/gnmi1/gnmi_pb2.py
@@ -26,3 +26,3 @@ from google.protobuf import any_pb2 as google_dot_protobuf_dot_any__pb2
 from google.protobuf import descriptor_pb2 as google_dot_protobuf_dot_descriptor__pb2
-from gnmi1 import gnmi_ext_pb2 as gnmi1_dot_gnmi__ext__pb2
+from . import gnmi_ext_pb2 as gnmi1_dot_gnmi__ext__pb2
 
diff --git b/finsy/proto/gnmi1/gnmi_pb2.pyi a/finsy/proto/gnmi1/gnmi_pb2.pyi
index b5f5883..f0e2f3c 100644
--- b/finsy/proto/gnmi1/gnmi_pb2.pyi
+++ a/finsy/proto/gnmi1/gnmi_pb2.pyi
@@ -14,3 +14,3 @@ http://github.com/openconfig/reference/blob/master/rpc/gnmi
 from collections import abc as _abc
-from gnmi1 import gnmi_ext_pb2 as _gnmi_ext_pb2
+from . import gnmi_ext_pb2 as _gnmi_ext_pb2
 from google.protobuf import any_pb2 as _any_pb2
diff --git b/finsy/proto/gnmi1/gnmi_pb2_grpc.py a/finsy/proto/gnmi1/gnmi_pb2_grpc.py
index 2b06c7a..9261aa5 100644
--- b/finsy/proto/gnmi1/gnmi_pb2_grpc.py
+++ a/finsy/proto/gnmi1/gnmi_pb2_grpc.py
@@ -5,3 +5,3 @@ import warnings
 
-from gnmi1 import gnmi_pb2 as gnmi1_dot_gnmi__pb2
+from . import gnmi_pb2 as gnmi1_dot_gnmi__pb2
 
diff --git b/finsy/proto/gnmi1/gnmi_pb2_grpc.pyi a/finsy/proto/gnmi1/gnmi_pb2_grpc.pyi
index f7fa75c..f189276 100644
--- b/finsy/proto/gnmi1/gnmi_pb2_grpc.pyi
+++ a/finsy/proto/gnmi1/gnmi_pb2_grpc.pyi
@@ -14,3 +14,3 @@ http://github.com/openconfig/reference/blob/master/rpc/gnmi
 from collections import abc as _abc
-from gnmi1 import gnmi_pb2 as _gnmi_pb2
+from . import gnmi_pb2 as _gnmi_pb2
 from grpc import aio as _aio
diff --git b/finsy/proto/p4/config/v1/p4info_pb2.py a/finsy/proto/p4/config/v1/p4info_pb2.py
index d5d4397..989e86c 100644
--- b/finsy/proto/p4/config/v1/p4info_pb2.py
+++ a/finsy/proto/p4/config/v1/p4info_pb2.py
@@ -25,3 +25,3 @@ _sym_db = _symbol_database.Default()
 from google.protobuf import any_pb2 as google_dot_protobuf_dot_any__pb2
-from p4.config.v1 import p4types_pb2 as p4_dot_config_dot_v1_dot_p4types__pb2
+from . import p4types_pb2 as p4_dot_config_dot_v1_dot_p4types__pb2
 
diff --git b/finsy/proto/p4/config/v1/p4info_pb2.pyi a/finsy/proto/p4/config/v1/p4info_pb2.pyi
index 451a295..3bf9464 100644
--- b/finsy/proto/p4/config/v1/p4info_pb2.pyi
+++ a/finsy/proto/p4/config/v1/p4info_pb2.pyi
@@ -11,3 +11,3 @@ from google.protobuf.internal import containers as _containers
 from google.protobuf.internal import enum_type_wrapper as _enum_type_wrapper
-from p4.config.v1 import p4types_pb2 as _p4types_pb2
+from . import p4types_pb2 as _p4types_pb2
 import builtins as _builtins
diff --git b/finsy/proto/p4/v1/p4runtime_pb2.py a/finsy/proto/p4/v1/p4runtime_pb2.py
index 96187af..ac43f11 100644
--- b/finsy/proto/p4/v1/p4runtime_pb2.py
+++ a/finsy/proto/p4/v1/p4runtime_pb2.py
@@ -25,5 +25,5 @@ _sym_db = _symbol_database.Default()
 from google.protobuf import any_pb2 as google_dot_protobuf_dot_any__pb2
-from google.rpc import status_pb2 as google_dot_rpc_dot_status__pb2
-from p4.config.v1 import p4info_pb2 as p4_dot_config_dot_v1_dot_p4info__pb2
-from p4.v1 import p4data_pb2 as p4_dot_v1_dot_p4data__pb2
+from ...google.rpc import status_pb2 as google_dot_rpc_dot_status__pb2
+from ..config.v1 import p4info_pb2 as p4_dot_config_dot_v1_dot_p4info__pb2
+from . import p4data_pb2 as p4_dot_v1_dot_p4data__pb2
 
diff --git b/finsy/proto/p4/v1/p4runtime_pb2.pyi a/finsy/proto/p4/v1/p4runtime_pb2.pyi
index c425d0b..d8c0746 100644
--- b/finsy/proto/p4/v1/p4runtime_pb2.pyi
+++ a/finsy/proto/p4/v1/p4runtime_pb2.pyi
@@ -11,5 +11,5 @@ from google.protobuf.internal import containers as _containers
 from google.protobuf.internal import enum_type_wrapper as _enum_type_wrapper
-from google.rpc import status_pb2 as _status_pb2
-from p4.config.v1 import p4info_pb2 as _p4info_pb2
-from p4.v1 import p4data_pb2 as _p4data_pb2
+from ...google.rpc import status_pb2 as _status_pb2
+from ..config.v1 import p4info_pb2 as _p4info_pb2
+from . import p4data_pb2 as _p4data_pb2
 import builtins as _builtins
diff --git b/finsy/proto/p4/v1/p4runtime_pb2_grpc.py a/finsy/proto/p4/v1/p4runtime_pb2_grpc.py
index dfff31c..940e083 100644
--- b/finsy/proto/p4/v1/p4runtime_pb2_grpc.py
+++ a/finsy/proto/p4/v1/p4runtime_pb2_grpc.py
@@ -5,3 +5,3 @@ import warnings
 
-from p4.v1 import p4runtime_pb2 as p4_dot_v1_dot_p4runtime__pb2
+from . import p4runtime_pb2 as p4_dot_v1_dot_p4runtime__pb2
 
diff --git b/finsy/proto/p4/v1/p4runtime_pb2_grpc.pyi a/finsy/proto/p4/v1/p4runtime_pb2_grpc.pyi
index 2b04e8a..b71fa56 100644
--- b/finsy/proto/p4/v1/p4runtime_pb2_grpc.pyi
+++ a/finsy/proto/p4/v1/p4runtime_pb2_grpc.pyi
@@ -7,3 +7,3 @@ from collections import abc as _abc
 from grpc import aio as _aio
-from p4.v1 import p4runtime_pb2 as _p4runtime_pb2
+from . import p4runtime_pb2 as _p4runtime_pb2
 import abc as _abc_1
diff --git b/finsy/proto/p4testgen1/p4testgen_pb2.py a/finsy/proto/p4testgen1/p4testgen_pb2.py
index 7032aca..1bf7c66 100644
--- b/finsy/proto/p4testgen1/p4testgen_pb2.py
+++ a/finsy/proto/p4testgen1/p4testgen_pb2.py
@@ -24,3 +24,3 @@ _sym_db = _symbol_database.Default()
 
-from p4.v1 import p4runtime_pb2 as p4_dot_v1_dot_p4runtime__pb2
+from ..p4.v1 import p4runtime_pb2 as p4_dot_v1_dot_p4runtime__pb2
 
diff --git b/finsy/proto/p4testgen1/p4testgen_pb2.pyi a/finsy/proto/p4testgen1/p4testgen_pb2.pyi
index 328503d..ca0baaf 100644
--- b/finsy/proto/p4testgen1/p4testgen_pb2.pyi
+++ a/finsy/proto/p4testgen1/p4testgen_pb2.pyi
@@ -9,3 +9,3 @@ from google.protobuf import message as _message
 from google.protobuf.internal import containers as _containers
-from p4.v1 import p4runtime_pb2 as _p4runtime_pb2
+from ..p4.v1 import p4runtime_pb2 as _p4runtime_pb2
 import builtins as _builtins
